### PR TITLE
Issue 6406: Pravega CLI commands that allow updating values in Controller Metadata tables

### LINKED ISSUE #6406: Pravega CLI commands that allow updating values in Controller Metadata tables
**Is your feature request related to a problem? Please describe.**
The Pravega CLI should provide commands that allow an admin to update the value corresponding to a specific key in a Controller metadata table.
This should help remedy any unexpected issues with Controller metadata in a production environment.

**Describe the solution you'd like**
Update command for updating any Controller Metadata table could look like:
```
controller-metadata put <table-name> <key> <ValueRecord.subfeild.subfeild...> <value> <segment-store endpoint>
```
For example, to update the Stream State in StreamMetadata Table we'd issue a command like:
```
controller-metadata put  _system/_tables/testScope/testStream/metadata.#.a72083e1-3ee1-468d-8d33-a61bef474441> "state"  StateRecord.state  "ACTIVE" <segment-store endpoint>
```

### PR DESCRIPTION
**Pull request by anirudhkovuru**
**Change log description**  
Implements the following command:
- `controller-metadata update <table-name> <key> <new-value-json-file> <segmentstore-endpoint>`

An update to an existing one:
- `controller-metadata get <table-name> <key> <[OPTIONAL] json-file> <segmentstore-endpoint>`

**Purpose of the change**  
Fixes #6406

**What the code does**  
***Commands***
- For the `update` command.
    - Added a new class for the `update` command.
    - The command attempts a conditional update by querying the current version and then performing the update to the table using that.
    - The new value to be updated in the table is provided as a JSON file of the record.
    - The necessary segmentstore calls are made through the `AdminSegmentHelper`.
- For the `get` command.
    - The `get` command is now updated to have an optional file argument. If given, the queried record is stored in JSON format at the provided location, else the queried record is printed to the console in a user friendly manner.
    - This is to make sure that the user can generate JSON's to be used in the `update`. The user only need to tweak the JSON and pass it into `update`.

***JSON serializers***
- Added a new class `ControllerMetadataJsonSerializer` to deal with converting the records to and from JSON.
- This contains internal classes used as `TypeAdaptor`s to help deal with specific data types such as `ImmutableMap`, `ImmutableList`, `ImmutableSet` and `Map.Entry<Double, Double>`.

***Minor changes***
- The `createFileAndDirectory` method was moved from the `ReadSegmentRangeCommand` class to its own tested `FileHelper` class for general use.

**How to verify it**  
All tests must pass.

### CODE REVIEW

**Comment by anirudhkovuru** [comment]
Need to add documentation.

File: cli/admin/README.md [hunk 1]
```diff
@@ -97,27 +97,28 @@ All available commands:
 	cluster list-instances : Lists all nodes in the Pravega cluster (Controllers, Segment Stores).
 	config list : Lists all configuration set during this session.
 	config set <name=value list>: Sets one or more config values for use during this session.
+	container continuous-recover <number-of-runs> <seconds-between-runs>: Executes a local, non-invasive recovery for all SegmentContainers in the cluster during the specified duration.
```
Comment (pbelgundi) [inline cli/admin/README.md:100]: Was this missed documentation?

Comment (anirudhkovuru) [reply cli/admin/README.md:100]: No, the order of the commands might've been mixed around as I updated the output of the `help` command in the documentation to reflect the current list of commands.

File: cli/admin/README.md [hunk 2]
```diff
@@ -97,27 +97,28 @@ All available commands:
...
 	container recover <container-id>: Executes a local, non-invasive recovery for a SegmentContainer.
-	container continuous-recover <container-id>: Executes a local, non-invasive recovery for all SegmentContainers in the cluster during the specified duration.
 	controller describe-readergroup <scope-name> <readergroup-id>: Get the details of a given ReaderGroup in a Scope.
 	controller describe-scope <scope-name>: Get the details of a given Scope.
 	controller describe-stream <scope-name> <stream-name>: Get the details of a given Stream.
 	controller list-readergroups <scope-name>: Lists all the existing ReaderGroups in a given Scope.
 	controller list-scopes : Lists all the existing scopes in the system.
 	controller list-streams <scope-name>: Lists all the existing Streams in a given Scope.
-	controller-metadata get <qualified-table-segment-name> <key> <segmentstore-endpoint>: Get the controller metadata entry for the given key in the table.
+	controller-metadata get <qualified-table-segment-name> <key> <json-file[OPTIONAL]> <segmentstore-endpoint>: Get the controller metadata entry for the given key in the table.
```
Comment (pbelgundi) [inline cli/admin/README.md:109]: All optional parameters should always be added after mandatory parameters to a command, also it would be good to add a flag like `--json` to indicate the user requires the output in json format

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/ReadSegmentRangeCommand.java [hunk 3]
```diff
@@ -66,27 +67,6 @@ public void execute() throws Exception {
         output("\nThe segment data has been successfully written into %s", fileName);
     }
 
-    /**
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/ReadSegmentRangeCommand.java:69]: This seems like is moved to the FileHelper class.

**Review by pbelgundi — CHANGES_REQUESTED** [review]
Mostly looks good except for a few minor comments.
Please also update output of manual test results here.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/ReadSegmentRangeCommand.java (hunk 3 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/ReadSegmentRangeCommand.java:69]: Yes this has been moved to the `FileHelper` class.

**Comment by anirudhkovuru** [comment]
The commands have been tested against the standalone.
```
> controller-metadata get _system/_tables/testScope/testStream/metadata.#.e81e4540-fc82-443d-b04e-fc9b073eaa97 configuration localhost test/config.json
For the given key: configuration
Successfully wrote the value to test/config.json in JSON.
```

The file `test/config.json` contains
```
{
  "scope": "testScope",
  "streamName": "testStream",
  "streamConfiguration": {
    "scalingPolicy": {
      "scaleType": "FIXED_NUM_SEGMENTS",
      "targetRate": 0,
      "scaleFactor": 0,
      "minNumSegments": 4
    },
    "retentionPolicy": null,
    "timestampAggregationTimeout": 0,
    "tags": [],
    "rolloverSizeBytes": 0
  },
  "updating": false,
  "tagOnlyUpdate": false,
  "removeTags": []
}
```

In the file `minNumSegments` is changed to 10.
```
{
  "scope": "testScope",
  "streamName": "testStream",
  "streamConfiguration": {
    "scalingPolicy": {
      "scaleType": "FIXED_NUM_SEGMENTS",
      "targetRate": 0,
      "scaleFactor": 0,
      "minNumSegments": 10
    },
    "retentionPolicy": null,
    "timestampAggregationTimeout": 0,
    "tags": [],
    "rolloverSizeBytes": 0
  },
  "updating": false,
  "tagOnlyUpdate": false,
  "removeTags": []
}
```

We then run the `update` command:
```
> controller-metadata update _system/_tables/testScope/testStream/metadata.#.e81e4540-fc82-443d-b04e-fc9b073eaa97 configuration test/config.json localhost
Successfully updated the key configuration in table _system/_tables/testScope/testStream/metadata.#.e81e4540-fc82-443d-b04e-fc9b073eaa97 with version 1020

> controller-metadata get _system/_tables/testScope/testStream/metadata.#.e81e4540-fc82-443d-b04e-fc9b073eaa97 configuration localhost
For the given key: configuration
StreamConfigurationRecord metadata info: 
scope = testScope
streamName = testStream
streamConfiguration = 
    scalingPolicy = ScalingPolicy(scaleType=FIXED_NUM_SEGMENTS, targetRate=0, scaleFactor=0, minNumSegments=10)
    retentionPolicy = null
    timestampAggregationTimeout = 0
    tags = []
    rolloverSizeBytes = 0
updating = false
tagOnlyUpdate = false
removeTags = []
```

File: cli/admin/README.md [hunk 4]
```diff
@@ -97,27 +97,28 @@ All available commands:
...
 	container recover <container-id>: Executes a local, non-invasive recovery for a SegmentContainer.
-	container continuous-recover <container-id>: Executes a local, non-invasive recovery for all SegmentContainers in the cluster during the specified duration.
 	controller describe-readergroup <scope-name> <readergroup-id>: Get the details of a given ReaderGroup in a Scope.
 	controller describe-scope <scope-name>: Get the details of a given Scope.
 	controller describe-stream <scope-name> <stream-name>: Get the details of a given Stream.
 	controller list-readergroups <scope-name>: Lists all the existing ReaderGroups in a given Scope.
 	controller list-scopes : Lists all the existing scopes in the system.
 	controller list-streams <scope-name>: Lists all the existing Streams in a given Scope.
-	controller-metadata get <qualified-table-segment-name> <key> <segmentstore-endpoint>: Get the controller metadata entry for the given key in the table.
+	controller-metadata get <qualified-table-segment-name> <key> <segmentstore-endpoint> <json-file[OPTIONAL]>: Get the controller metadata entry for the given key in the table.
```
Comment (pbelgundi) [inline cli/admin/README.md:109]: Description could perhaps be :
`Get the value for the specified key from the specified ControllerMetadata Table.`

A common convention for specifying optional arguments to a command is to enclose them in square brackets (this is followed by almost all unix and windows shells). So your command documentation should look like:
```
controller-metadata get <qualified-table-segment-name> <key> <segmentstore-endpoint> [json-file]
```

File: cli/admin/README.md [hunk 5]
```diff
@@ -97,27 +97,28 @@ All available commands:
...
 	controller-metadata list-entries <qualified-table-segment-name> <entry-count> <segmentstore-endpoint>: List at most the required number of entries from the controller metadata table. Unsupported for stream metadata tables.
 	controller-metadata list-keys <qualified-table-segment-name> <key-count> <segmentstore-endpoint>: List at most the required number of keys from the controller metadata table.
 	controller-metadata tables-info : List all the controller metadata tables.
+	controller-metadata update <qualified-table-segment-name> <key> <new-value-file> <segmentstore-endpoint>: Update the given key in the table with the provided value.
```
Comment (pbelgundi) [inline cli/admin/README.md:113]: Try to keep out such formatting changes from your PR. That makes very hard to review...

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: cli/admin/README.md (hunk 1 above)
Comment (pbelgundi) [reply cli/admin/README.md:100]: ok

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        super(args);
+    }
+
+    @Override
+    public void execute() throws Exception {
+        ensureArgCount(4);
+
+        final String tableName = getArg(0);
+        final String key = getArg(1);
+        final String newValueFile = getArg(2);
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java:49]: This does not look right. I thought you made the json file the last argument to the update command. Please test update command after this change as well.

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        ControllerMetadataJsonSerializer jsonSerializer = new ControllerMetadataJsonSerializer();
+
+        String jsonValue;
+        try {
+            jsonValue = new String(Files.readAllBytes(Paths.get(newValueFile)));
+        } catch (NoSuchFileException e) {
+            output("File with new value does not exist: %s", newValueFile);
+            return;
+        }
+        ByteBuffer updatedValue = serializer.serialize(jsonSerializer.fromJson(jsonValue, serializer.getMetadataClass()));
```
Comment (pbelgundi) [inline cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java:65]: Please also test for negative test scenarios where the jsonValue supplied by the User is malformed or not as per expectation... I would expect to see a JsonParseException thrown at this line... but it would be good to test and verify.

**Review by pbelgundi — CHANGES_REQUESTED** [review]

File: cli/admin/README.md (hunk 4 above)
Comment (anirudhkovuru) [reply cli/admin/README.md]: Changed it accordingly.

File: cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java (hunk 7 above)
Comment (anirudhkovuru) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java:65]: I've added a test for the any JSON exception cases.

Comment (pbelgundi) [reply cli/admin/src/main/java/io/pravega/cli/admin/controller/metadata/ControllerMetadataUpdateEntryCommand.java:65]: Cool Thanks.

**Review by pbelgundi — APPROVED** [review]
LGTM

**Comment by pbelgundi** [comment]
tests have passed on branch builder: pravega-branch-builder/1800
